### PR TITLE
fix(gnovm): follow up works on loop scope(#2429)

### PR DESCRIPTION
**Pull request by ltzmaxwell**
- Complete the remaining work from #2429.
- Go 1.22 loopvar are not fully supported. (should be supported in next Version).
   e.g.  
   this is supported,

```go
        for i, _ := range s {
		s1 = append(s1, &i)
	}
```
   
this is not supported:

```go
        for i := 0; i < 3; i++ {
		fs = append(fs, func() { println(i) })
	}
```

### CODE REVIEW

File: gnovm/pkg/gnolang/preprocess.go [hunk 1]
```diff
@@ -2291,6 +2291,21 @@ func findGotoLoopDefines(ctx BlockNode, bn BlockNode) {
...
+							switch n := n.(type) {
+							case *NameExpr:
+								if n.Type == NameExprTypeDefine {
+									n.Type = NameExprTypeHeapDefine
+								}
+							}
+						}
+						return n, TRANS_CONTINUE
+					})
+
```
Comment (jaekwon) [inline gnovm/pkg/gnolang/preprocess.go:2308]: this is missing a lot of the logic that is needed, which is present in the BranchStmt's Transcribe.
TRANS_ENTER *FuncLitExpr should return SKIP similar, because LoopDefines should be defined within the for-loop but not within a containing closure.

Comment (ltzmaxwell) [reply gnovm/pkg/gnolang/preprocess.go]: fixed.

File: gnovm/pkg/gnolang/preprocess.go [hunk 2]
```diff
@@ -222,7 +222,7 @@ func initStaticBlocks(store Store, ctx BlockNode, bn BlockNode) {
 				}
 				for i := range n.Results {
 					r := &n.Results[i]
-					if r.Name == blankIdentifier {
+					if r.Name == blankIdentifier || r.Name == "" {
```
Comment (ltzmaxwell) [inline gnovm/pkg/gnolang/preprocess.go:225]: keep(restore to) the original init order.

File: gno.land/pkg/sdk/vm/gas_test.go [hunk 3]
```diff
@@ -67,7 +67,7 @@ func TestAddPkgDeliverTx(t *testing.T) {
 	gasDeliver := gctx.GasMeter().GasConsumed()
 
 	assert.True(t, res.IsOK())
-	assert.Equal(t, int64(87809), gasDeliver)
+	assert.Equal(t, int64(88289), gasDeliver)
```
Comment (ltzmaxwell) [inline gno.land/pkg/sdk/vm/gas_test.go:70]: Side effect caused by initializing res_n in initStaticBlock.

File: gnovm/pkg/gnolang/preprocess.go [hunk 4]
```diff
@@ -296,10 +303,12 @@ func initStaticBlocks(store Store, ctx BlockNode, bn BlockNode) {
 					}
 					n.Predefine(false, pte.Name)
 				}
-				for _, rte := range n.Type.Results {
-					if rte.Name != "" {
-						n.Predefine(false, rte.Name)
+				for i, rte := range n.Type.Results {
+					if rte.Name == "" {
+						rn := fmt.Sprintf(".res_%d", i)
+						rte.Name = Name(rn)
```
Comment (ltzmaxwell) [inline gnovm/pkg/gnolang/preprocess.go:309]: define in `initStaticBlocks` phase to preserve original order, which will impact defer 
behavior.

File: gnovm/pkg/gnolang/preprocess.go [hunk 5]
```diff
@@ -296,10 +303,12 @@ func initStaticBlocks(store Store, ctx BlockNode, bn BlockNode) {
...
 					}
+					n.Predefine(false, rte.Name)
```
Comment (ltzmaxwell) [inline gnovm/pkg/gnolang/preprocess.go:311]: define in initStaticBlocks phase to preserve original order, which will impact defer behavior.

File: examples/gno.land/p/demo/avl/z_0_filetest.gno [hunk 6]
```diff
@@ -246,11 +265,11 @@ func main() {
 //                 "Closure": {
 //                     "@type": "/gno.RefValue",
 //                     "Escaped": true,
-//                     "ObjectID": "a8ada09dee16d791fd406d629fe29bb0ed084a30:3"
+//                     "ObjectID": "a8ada09dee16d791fd406d629fe29bb0ed084a30:5"
 //                 },
 //                 "FileName": "main.gno",
 //                 "IsMethod": false,
-//                 "Name": "init.0",
+//                 "Name": "init.1",
```
Comment (ltzmaxwell) [inline examples/gno.land/p/demo/avl/z_0_filetest.gno:272]: is this introduced by the `initStaticBlocks`?

File: examples/gno.land/p/demo/avl/z_0_filetest.gno [hunk 7]
```diff
@@ -237,6 +237,25 @@ func main() {
 //     "Values": [
 //         {
 //             "T": {
+//                 "@type": "/gno.PointerType",
```
Comment (ltzmaxwell) [inline examples/gno.land/p/demo/avl/z_0_filetest.gno:240]: ditto

File: examples/gno.land/p/demo/avl/z_0_filetest.gno [hunk 8]
```diff
@@ -237,6 +237,25 @@ func main() {
...
+//                     "Hash": "e925bb6bf884ece9620dd27f831dc510b5bc0e55",
+//                     "ObjectID": "a8ada09dee16d791fd406d629fe29bb0ed084a30:6"
+//                 },
+//                 "Index": "0",
+//                 "TV": null
+//             }
+//         },
+//         {
+//             "T": {
 //                 "@type": "/gno.FuncType",
```
Comment (ltzmaxwell) [inline examples/gno.land/p/demo/avl/z_0_filetest.gno:259]: ditto

Comment (ltzmaxwell) [reply examples/gno.land/p/demo/avl/z_0_filetest.gno:259]: package var is inited firstly in `initStaticBlocks`, the the order changes.

File: examples/gno.land/p/demo/avl/z_0_filetest.gno (hunk 6 above)
Comment (ltzmaxwell) [reply examples/gno.land/p/demo/avl/z_0_filetest.gno:272]: yes.

File: gnovm/pkg/gnolang/preprocess.go [hunk 9]
```diff
@@ -4311,7 +4366,7 @@ func setNodeLocations(pkgPath string, fileName string, n Node) {
 // XXX check node lines, uniqueness of locations,
 // and also check location pkgpath and filename.
 func checkNodeLinesLocations(pkgPath string, fileName string, n Node) {
-	// XXX
+	// TODO: XXX
```
Comment (ltzmaxwell) [inline gnovm/pkg/gnolang/preprocess.go:4369]: should be in another PR.

File: gnovm/pkg/gnolang/preprocess.go [hunk 10]
```diff
@@ -2667,36 +2718,21 @@ func findLoopUses2(ctx BlockNode, bn BlockNode) {
...
 				if len(lds) < len(lus) {
 					panic("defines should be a superset of used-defines")
 				}
-				n.SetAttribute(ATTR_LOOP_DEFINES, lus)
-				// XXX implement delete.
-				// n.DelAttribute(ATTR_LOOP_USES)
+				// no need anymore
+				n.DelAttribute(ATTR_LOOP_USES)
+				n.DelAttribute(ATTR_LOOP_DEFINES)
 			}
```
Comment (ltzmaxwell) [inline gnovm/pkg/gnolang/preprocess.go:2729]: no need anymore?

File: gnovm/pkg/gnolang/preprocess.go [hunk 11]
```diff
@@ -247,10 +246,18 @@ func initStaticBlocks(store Store, ctx BlockNode, bn BlockNode) {
 			case *RangeStmt:
 				if n.Op == DEFINE {
 					if n.Key != nil {
-						last.Predefine(false, n.Key.(*NameExpr).Name)
+						nx := n.Key.(*NameExpr)
+						if nx.Name != blankIdentifier {
+							nx.Type = NameExprTypeDefine
+							last.Predefine(false, nx.Name)
+						}
 					}
```
Comment (ltzmaxwell) [inline gnovm/pkg/gnolang/preprocess.go:254]: maybe we should not support loopvar with range either, to make it consistent with for loop.

Comment (jaekwon) [reply gnovm/pkg/gnolang/preprocess.go:254]: yes. but TypeDefine is correct here as it is a define.

File: gnovm/pkg/gnolang/preprocess.go [hunk 12]
```diff
@@ -2549,28 +2569,43 @@ func addHeapCapture(dbn BlockNode, fle *FuncLitExpr, name Name) uint16 {
 	ne := NameExpr{
 		Path: vp,
 		Name: name,
-		Type: NameExprTypeHeapClosure,
+		Type: NameExprTypeHeapClosure, // XXX, this is actually not used, remove?
```
Comment (jaekwon) [inline gnovm/pkg/gnolang/preprocess.go:2572]: right it's not actually used, but let's leave it because it is a special kind of use of a name, to copy it to func value captures. It just means it was captured by a closure, for now.

File: gnovm/pkg/gnolang/op_call.go [hunk 13]
```diff
@@ -57,6 +57,19 @@ func (m *Machine) doOpCall() {
 	// Create new block scope.
 	clo := fr.Func.GetClosure(m.Store)
 	b := m.Alloc.NewBlock(fr.Func.GetSource(m.Store), clo)
+
+	// Copy *FuncValue.Captures into block
+	// NOTE: addHeapCapture in preprocess ensures order.
+	// XXX, actually copy
```
Comment (jaekwon) [inline gnovm/pkg/gnolang/op_call.go:63]: isn't it actually copying? why XXX?

**Comment by jaekwon** [comment]
Brilliant.
